VDA-SBT/contracts/SoulboundNFT.sol

```diff
@@ -0,0 +1,316 @@
+            return false;
+        }
+        return true;
+    }
+
+    /**
+     * @dev See {ISoulboundNFT}
+     */
+    function totalSupply() external view onlyOwner override returns(uint){
+        return _tokenIdCounter.current();
+    }
+
+    /**
+     * @dev See {ISoulboundNFT}
+     */
+    function addCompanyAccount(address account) external onlyOwner override {
+        require(!_companyAccounts.contains(account), "Existing account");
+        _companyAccounts.add(account);
+
+        emit AddCompanyAccount(account);
+    }
+
+    /**
+     * @dev See {ISoulboundNFT}
+     */
+    function removeCompanyAccount(address account) external onlyOwner override {
+        require(_companyAccounts.contains(account), "Invalid account");
+        _companyAccounts.remove(account);
+
+        emit RemoveCompanyAccount(account);
+    }
+
+    /**
+     * @dev See {ISoulboundNFT}
+     */
+    function listCompanyAccounts() external view override returns(address[] memory) {
+        uint length = _companyAccounts.length();
+        address[] memory list = new address[](length);
+        for (uint i = 0; i < length; i++) {
+            list[i] = _companyAccounts.at(i);
+        }
+
+        return list;
+    }
+
+    /**
+     * @dev See {ISoulboundNFT}
+     */
+    function addSBTType(string calldata sbtType) external onlyOwner override {
+        require(!_sbtTypes.contains(sbtType), "Existing SBT type");
+        _sbtTypes.add(sbtType);
+
+        emit AddSBTType(sbtType);
+    }
+
+    /**
+     * @dev See {ISoulboundNFT}
+     */
+    /*
+    function removeSBTType(string calldata sbtType) external onlyOwner override {
+        require(_sbtTypes.contains(sbtType), "Not registered type");
+        _sbtTypes.remove(sbtType);
+
+        emit RemoveSBTType(sbtType);
+    }
+    */
+
+    /**
+     * @dev See {ISoulboundNFT}
+     */
+    function allowedSBTTypes() external view override returns(string[] memory) {
+        // To-do
+        uint length = _sbtTypes.length();
+
+        string[] memory sbtTypeList = new string[](length);
+
+        for (uint i = 0; i < length; i++) {
+            sbtTypeList[i] = _sbtTypes.at(i);
+        }
+
+        return sbtTypeList;
+    }
+
+    /**
+     * @dev See {IsoulboundNFT}
+     */
+    function claimSBT(
+        address did,
+        string calldata sbtType,
+        string calldata uniqueId,
+        bytes calldata signature,
+        bytes calldata proof
+    ) external onlyValidSBTType(sbtType) override returns(uint) {
+        require(_sbtInfo[sbtType][msg.sender] == 0, "Already claimed type");
+        
+        {
+            address stackDID = did;
+            uint didNonce = getNonce(did);
+            bytes memory params = abi.encodePacked(
+                sbtType,
+                "-",
+                uniqueId,
+                "-",
+                stackDID,
+                "-",
+                didNonce
+            );
+            verifyRequest(did, params, signature, proof);
```

Comment:
@todo This verification code will need to use the Verida Verification Base Contract when it's ready.

Reply:
SBT contract can't use the VerificationBaseContract's verifyRequest() function.
Because in the VerificationBaseContract, verifyRequest() function verify the signer is the `did` in the input parameters.
But in the SBT contract, we should iterate all the company accounts to verify the signature is signed by one of registered company accounts.

Reply:
So, in the SouboundNFT contract, I overrided the `verifyRequest(...)` function.

Reply:
Please check the verifyRequest() function in the VerificationBaseContract. 
![image](https://user-images.githubusercontent.com/15656252/203179005-b6c243c0-d137-42bb-950f-276dde759278.png)
Once the recovered address doesn't equal to the inputed `did`, it will reject the entire transaction.
In our case, we should iterate through all our company address, and check one of them is matched.
So, we can't use the original verifyRequest function.

Reply:
Yes. I think we should change verifyRequest to provide an array if `dids`:

`verifyRequest(address[] dids, ...)`

and then 

`require(proofSigner in dids)`

(pseudocode)

Reply:
Means we should update verifyRequest() function in the VerificationBase contract?